### DISCUSSION TITLE
Write to slave (change address ID)

### DISCUSSION THREAD

**Discussion by imojams** [eModbus/eModbus#139] [Q&A] [answered]
Hello.

I try to change slave address of energy meter [OR-WE-514](https://orno.pl/en/search?type=products&phrase=or-we-514) 

Modbus register list in [PDF](https://b2b.orno.pl/download-resource/26060/)

From PDF:
system parameters address area: 03H read, 10H, 06H write function code, low address is high byte data.
So, 06H = WRITE_HOLD_REGISTER

Request: MB.addRequest(100, 1, WRITE_HOLD_REGISTER, 0x0110, 0x15); // 21
But i got error: Error response: 01 - Illegal function code

If I issue addres read request, it prints current address - 1 (default);
MB.addRequest(102, 1, READ_HOLD_REGISTER, 0x0110, 1); // 21

How to correctly send WRITE_HOLD_REGISTER command?

**Comment by Miq1** [comment]
Hi,

your WRITE_HOLD_REGISTER call is perfectly correct. The 01 return code must have come from the meter, although the documentation you provided tells us the register shall be writable.

You could enable logging by adding ``-DLOG_LEVEL=6`` to your compiler options. Then we would see in the Serial log what messages went back and forth.

**Comment by imojams** [comment]
Here is output from serial monitor:

21:11:45.296 -> [V] 63067| ModbusMessage.cpp    [ 529] checkData: Check data #3
21:11:45.296 -> [D] 63067| ModbusClientRTU.cpp  [ 107] addRequestM: request for 01/06
21:11:45.296 -> [D] 63077| ModbusClientRTU.cpp  [ 156] addToQueue: RC=01
21:11:45.296 -> [D] 63077| ModbusClientRTU.cpp  [ 118] addRequestM: RC=00
21:11:45.296 -> Address change request sent! Waiting one minute ...
21:11:45.296 -> [D] 63077| ModbusClientRTU.cpp  [ 174] handleConnection: Pulled request from queue
21:11:45.343 -> [D] Sent packet: @3FFB9904/6:
21:11:45.343 ->   | 0000: 01 06 01 10 00 15                                 |......          |
21:11:45.343 -> [D] 63107| ModbusClientRTU.cpp  [ 179] handleConnection: Request sent.
21:11:45.390 -> [D] Raw buffer received: @3FFB395C/5:
21:11:45.390 ->   | 0000: 01 86 01 83 A0                                    |.....           |
21:11:45.390 -> [D] Received packet: @3FFB9918/5:
21:11:45.390 ->   | 0000: 01 86 01 83 A0                                    |.....           |
21:11:45.390 -> [D] 63181| ModbusClientRTU.cpp  [ 185] handleConnection: Data response (5 bytes) received.
21:11:45.436 -> [V] Data: @3FFB9918/5:
21:11:45.436 ->   | 0000: 01 86 01 83 A0                                    |.....           |
21:11:45.436 -> [D] 63201| ModbusClientRTU.cpp  [ 214] handleConnection: Response generated.
21:11:45.436 -> [V] Response packet: @3FFB9918/3:
21:11:45.436 ->   | 0000: 01 86 01                                          |...             |
21:11:45.436 -> [1;33m[E] 63222| ModbusOrno514.ino    [ 118] handleModbusError: Error response: 01 - Illegal function code
21:11:45.436 -> [0m

**Reply by Miq1** [reply]
Nothing new, unfortunately. You will have to find out why the device is not accepting the function code. Could it be you will have to do something else in preparation for the server id change? That would be sensible to ptotect the device from being rendered inaccessible with a single request.

**Reply by imojams** [reply]
I will try manufacturer application and see if they have functionality to change ID.

**Comment by imojams** [comment, marked as answer]
With manufacturer app i can change modbus id: here is some output of app:

// Read modbus ID (modbus id = 1)
14:46:23:093  Over
14:46:33:067  Write: 01 03 01 10 00 01 84 33 
14:46:33:370  Read: 01 03 02 00 01 79 84 
14:46:33:686  Over

// write modbus id=21 (021)
14:46:51:818  Write: 01 10 01 10 00 01 02 00 15 75 CF 
14:46:52:120  Read: 01 10 01 10 00 01 01 F0 
14:46:52:442  Over

**Reply by Miq1** [reply]
Well, they are using a different function code: instead of WRITE_HOLD_REGISTER (0x06), they are using WRITE_MULT_REGISTERS (0x10).

**Reply by imojams** [reply]
Thanks for your time. 
Now meter is already installed. When i buy next one, then i will try multi write.

**Reply by Miq1** [reply]
No worries.

It is a bit silly/lazy not to add the 0x06 function code to the server's firmware, as it could be easily implemented by a thin wrapping function around the 0x10 code they do have implemented. Every fool is different, as we say.😁
